x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,229 @@
+            final Settings indexSettings = indexMetadata.getSettings();
+            if (SearchableSnapshotsSettings.isSearchableSnapshotStore(indexSettings)) {
+                final SnapshotId otherSnapshotId = new SnapshotId(
+                    SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexSettings),
+                    SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSettings)
+                );
+                if (Objects.equals(snapshotId, otherSnapshotId)) {
+                    final IndexId otherIndexId = new IndexId(
+                        SNAPSHOT_INDEX_NAME_SETTING.get(indexSettings),
+                        SNAPSHOT_INDEX_ID_SETTING.get(indexSettings)
+                    );
+                    if (Objects.equals(indexId, otherIndexId)) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    private class MaintenanceTask extends AbstractRunnable {
+
+        private final SnapshotId snapshotId;
+        private final IndexId indexId;
+        private final int numberOfShards;
+        private final Supplier<ClusterState> state;
+
+        MaintenanceTask(SnapshotId snapshotId, IndexId indexId, int numberOfShards, Supplier<ClusterState> clusterStateSupplier) {
+            this.snapshotId = Objects.requireNonNull(snapshotId);
+            this.indexId = Objects.requireNonNull(indexId);
+            this.numberOfShards = numberOfShards;
+            this.state = clusterStateSupplier;
+        }
+
+        @Override
+        protected void doRun() {
+            if (hasSearchableSnapshotWith(state.get(), snapshotId, indexId)) {
+                logger.debug(
+                    "snapshot blob cache maintenance task skipped, another index is using [snapshot:{}, index:{}]]",
+                    snapshotId,
+                    indexId
+                );
+                return;
+            }
+
+            final Set<String> paths = new HashSet<>(numberOfShards);
+            for (int shard = 0; shard < numberOfShards; shard++) {
+                paths.add(String.join("/", snapshotId.getUUID(), indexId.getId(), String.valueOf(shard)));
+            }
+
+            final DeleteByQueryRequest request = new DeleteByQueryRequest(systemIndexName);
+            request.setQuery(QueryBuilders.termsQuery("blob.path", paths));
+            clientWithOrigin.execute(DeleteByQueryAction.INSTANCE, request, new ActionListener<>() {
+                @Override
+                public void onResponse(BulkByScrollResponse response) {
+                    logger.debug(
+                        "snapshot blob cache maintenance task deleted [{}] documents for [snapshot:{}, index:{}]]",
+                        response.getDeleted(),
+                        snapshotId,
+                        indexId
+                    );
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.warn(
```

Comment:
I suppose we can expect this to happen rarely, i.e., we would still be leaking entries in the cache.

Reply:
Yes, we should be more explicit about leaked entries in the system index when logging. 

At the same time I think we should complete this on-index-deletion clean up with another clean up that would run every day (I think we have maintenance service like this) to clean up unused cache entries.